source/AST/ASTVisitor.cpp

```diff
@@ -1582,58 +1461,251 @@ TraverseFunctionTemplateDecl(
     // (e.g. for an abbreviated function template with no template-head)
     if(! shouldExtract(FD))
         return true;
+    auto Template = std::make_unique<TemplateInfo>();
+    parseTemplateParams(*Template, FD);
 
-    FunctionInfo I;
-    parseTemplateParams(I.Template, FD);
+    // traverse the templated declaration according to its kind
+    return TraverseDecl(FD, std::move(Template));
+}
 
-    switch(FD->getKind())
+bool
+ASTVisitor::
+Traverse(
+    ClassScopeFunctionSpecializationDecl* D)
+{
+    if(! shouldExtract(D))
+        return true;
+
+    // for class scope explicit specializations of member function templates which
+    // are members of class templates, it is impossible to know what the
+    // primary template is until the enclosing class template is instantiated.
+    // while such declarations are valid C++ (see CWG 727 and [temp.expl.spec] p3),
+    // GCC does not consider them to be valid. consequently, we do not extract the SymbolID
+    // of the primary template. in the future, we could take a best-effort approach to find
+    // the primary template, but this is only possible when none of the candidates are dependent
+    // upon a template parameter of the enclosing class template.
+    auto Template = std::make_unique<TemplateInfo>();
+    parseTemplateArgs(*Template, D);
+
+    CXXMethodDecl* MD = D->getSpecialization();
+
+    // KRYSTIAN FIXME: is the right? should this call TraverseDecl instead?
+    return Traverse(MD, std::move(Template));
+}
+
+bool
+ASTVisitor::
+Traverse(
+    TypeAliasTemplateDecl* D)
+{
+    TypeAliasDecl* AD = D->getTemplatedDecl();
+    if(! shouldExtract(AD))
+        return true;
+
+    auto Template = std::make_unique<TemplateInfo>();
+    parseTemplateParams(*Template, AD);
+
+    return Traverse(AD, std::move(Template));
+}
+
+template<typename... Args>
+auto
+ASTVisitor::
+Traverse(Args&&...)
```

Comment:
you could =delete this instead and save the unreachable theatrics.

You might even make it not a template and instead write
```
auto Traverse(...) = delete
```

Reply:
No can do, that would cause a hard error in `TraverseDecl`. The catch-all function has to be selectable by overload resolution without causing an error (a deleted definition makes the program ill-formed if it is referenced)

Reply:
boooo